In [3]:
import pandas as pd
import numpy as np


In [4]:

# 1. Create a small DataFrame with 3 rows and 3 columns
df_x = pd.DataFrame({
    'Feature_A': [1, 2, 3],
    'Feature_B': [10, 20, 30],
    'Feature_C': [100, 200, 300]
})

print(df_x)

   Feature_A  Feature_B  Feature_C
0          1         10        100
1          2         20        200
2          3         30        300


---

### 2. Applying `axis=1` vs `axis=0`

* **`axis=1` (Row-wise sum / Horizontally):**
It collapses columns together for each row.
* Row 0: $1 + 10 + 100 = \mathbf{111}$
* Row 1: $2 + 20 + 200 = \mathbf{222}$
* Row 2: $3 + 30 + 300 = \mathbf{333}$

In [5]:
    df_x.sum(axis=1)
# Output:
# 0    111
# 1    222
# 2    333

0    111
1    222
2    333
dtype: int64

In [6]:
# Create noise for 3 rows
np.random.seed(42)  # For reproducible results
noise = np.random.normal(loc=0, scale=1, size=3)
# e.g., noise might be [0.496, -0.138, 0.647]

y_true = df_x.sum(axis=1) + noise
print(y_true)

0    111.496714
1    221.861736
2    333.647689
dtype: float64


**Final Result (`y_true`):**

* Row 0: $111 + 0.496 = \mathbf{111.496}$
* Row 1: $222 - 0.138 = \mathbf{221.862}$
* Row 2: $333 + 0.647 = \mathbf{333.647}$

---

### Summary Rule of Thumb

* **`axis=1`**: Collapses **columns** $\rightarrow$ Moves **left-to-right** across each row.
* **`axis=0`**: Collapses **rows** $\rightarrow$ Moves **top-to-bottom** down each column.

This pattern (`y_true = df_x.sum(axis=1) + noise`) is commonly used in synthetic datasets to create a target variable $y$ that depends linearly on all input features in `df_x` plus some artificial measurement error.

In [7]:
import pandas as pd
import statsmodels.api as sm

df = pd.DataFrame({'A': [10, 20]})
df_with_const = sm.add_constant(df)

print(df_with_const['const'].values)
# Output: [1. 1.]

[1. 1.]


In [8]:
print(df_with_const)

   const   A
0    1.0  10
1    1.0  20


In [11]:

df_x = pd.DataFrame({
    'Feature_A': [1, 2, 3],
    'Feature_B': [10, 20, 30],
    'Feature_C': [100, 200, 300]
})
cols = df_x.columns

df_x_with_square = pd.concat((df_x, (df_x**2).rename(columns = {f"{var}": f"{var}2" for var in cols})), axis=1)
df_x_with_square

,Feature_A,Feature_B,Feature_C,Feature_A2,Feature_B2,Feature_C2
0,1,10,100,1,100,10000
1,2,20,200,4,400,40000
2,3,30,300,9,900,90000
